app/forms/application_form.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+class ApplicationForm
+  include ActiveModel::Conversion
+  include ActiveModel::Naming
+  include ActiveModel::Validations
+
+  # Overrides <tt>ActiveModel::Naming#model_name</tt> to take <tt>resource_name</tt>
+  # into account.
+  #
+  # @return [ActiveModel::Name]
+  def self.model_name
+    ActiveModel::Name.new(self, nil, resource_class.name)
+  end
+
+  # Sets or infers the resource name.
+  #
+  # @param resource [String] the name of the resource class
+  # @return [String]
+  # @example A class where the form object name does not match the resource:
+  #   class SignupForm < ApplicationForm
+  #     resource_class User
+  #   end
+  def self.resource_class(klass = nil)
+    @resource_class ||= (klass || name.chomp('Form').constantize)
+  end
+
+  # The underlying model wrapped by the form.
+  #
+  # @return [Type<ApplicationRecord>]
+  attr_accessor :resource
+  private :resource=
+
+  delegate :id, :persisted?, to: :resource
+
+  def initialize(resource, attributes = {})
```

Comment:
How can I use this form when there is no resource, for example `SignInForm`. I can pass something fake like `User.new` or `nil` but it seems that I'm doing something wrong. What do you think guys if we make two form objects?
```ruby
class ApplicationForm
  include ActiveModel::Model

  def save
    if valid?
      persist!
      true
    else
      false
    end
  end

  private

  def persist!; end
end

class ResourseForm < ApplicationForm
  # ...

  attr_accessor :resource
  private :resource=

  def initialize(resourse, attributes)
    super
    self.resourse = resourse
  end

  # ...
end
```
The first one is really basic one which we can use for simple form without resources (SignInForm) or we can use for very complicated form objects (creating/updating list of records). And second one is good for most of forms that just create or update simple records (UserCreateForm).

Reply:
Thanks for the feedback @vvmarulin. In fact I have a not yet pushed version of this that will default the resource to `resource_class.new`, which means you don't need to pass it in for `create` actions etc. 

As for your hypothetical `SigninForm`, how would that get away with not having a resource? You need the user or a similar object to get the stored password hash at least and if you don't your form may be too simple to need a form object anyway, no? 

Two base classes seems  a bit overkill for the rare case one doesn't need a resource. I guess I'd recommend just not inheriting from `ApplicationForm` at all in this case (if you don't need a PORO, you don't inherit from `ApplicationRecord`, so if you need a simpler form, don't inherit from `ApplicationForm`, that's the idea).